test/unit/cmap/handshake/client_metadata.test.ts

```diff
@@ -1,149 +1,511 @@
+            } finally {
+              cleanupEnv();
+            }
+          });
+        }
+      });
+    }
+
+    context('when a numeric FAAS env variable is not numerically parsable', () => {
+      before(() => {
+        process.env.AWS_EXECUTION_ENV = 'non-empty-string';
+        process.env.AWS_LAMBDA_FUNCTION_MEMORY_SIZE = '123not numeric';
+      });
+
+      after(() => {
+        delete process.env.AWS_EXECUTION_ENV;
+        delete process.env.AWS_LAMBDA_FUNCTION_MEMORY_SIZE;
+      });
+
+      it('does not attach it to the metadata', () => {
+        expect(makeClientMetadata({ driverInfo: {} })).not.to.have.nested.property('aws.memory_mb');
+      });
+    });
+  });
+
+  describe('metadata truncation', function () {
+    afterEach(() => sinon.restore());
+
+    context('when driverInfo is too large', () => {
+      it('throws an error relating to name', () => {
+        expect(() => makeClientMetadata({ driverInfo: { name: 'a'.repeat(512) } })).to.throw(
+          MongoInvalidArgumentError,
+          /name/
+        );
+      });
+
+      it('throws an error relating to version', () => {
+        expect(() => makeClientMetadata({ driverInfo: { version: 'a'.repeat(512) } })).to.throw(
+          MongoInvalidArgumentError,
+          /version/
+        );
+      });
+
+      it('throws an error relating to platform', () => {
+        expect(() => makeClientMetadata({ driverInfo: { platform: 'a'.repeat(512) } })).to.throw(
+          MongoInvalidArgumentError,
+          /platform/
+        );
+      });
+    });
+
+    context('when faas region is too large', () => {
+      beforeEach('1. Omit fields from `env` except `env.name`.', () => {
+        sinon.stub(process, 'env').get(() => ({
+          AWS_EXECUTION_ENV: 'iLoveJavaScript',
+          AWS_REGION: 'a'.repeat(512)
+        }));
+      });
+
+      it('only includes env.name', () => {
+        const metadata = makeClientMetadata({ driverInfo: {} });
+        expect(metadata).to.not.have.nested.property('env.region');
+        expect(metadata).to.have.nested.property('env.name', 'aws.lambda');
+        expect(metadata.env).to.have.all.keys('name');
+      });
+    });
+
+    context('when os information is too large', () => {
+      context('release too large', () => {
+        beforeEach('2. Omit fields from `os` except `os.type`.', () => {
+          sinon.stub(process, 'env').get(() => ({
+            AWS_EXECUTION_ENV: 'iLoveJavaScript',
+            AWS_REGION: 'abc'
+          }));
+          sinon.stub(os, 'release').returns('a'.repeat(512));
+        });
+
+        it('only includes env.name', () => {
+          const metadata = makeClientMetadata({ driverInfo: {} });
+          expect(metadata).to.have.property('env');
+          expect(metadata).to.have.nested.property('env.region', 'abc');
+          expect(metadata.os).to.have.all.keys('type');
+        });
+      });
+
+      context('os.type too large', () => {
```

Comment:
The scenario is likely never ever possible (I have to fake an OS with a 512 length name), but it demonstrates that we abide the strictness of the 512 bytes limit.